packages/textfield/src/mwc-textfield-base.ts

```diff
@@ -16,25 +16,42 @@ limitations under the License.
 */
 import '@material/mwc-notched-outline';
 
+import {MDCLineRippleFoundation} from '@material/line-ripple/foundation.js';
 import {addHasRemoveClass, classMap, FormElement, html, property, PropertyValues, query, TemplateResult} from '@material/mwc-base/form-element.js';
 import {floatingLabel, FloatingLabel} from '@material/mwc-floating-label';
 import {lineRipple, LineRipple} from '@material/mwc-line-ripple';
 import {NotchedOutline} from '@material/mwc-notched-outline';
 import {MDCTextFieldAdapter, MDCTextFieldInputAdapter, MDCTextFieldLabelAdapter, MDCTextFieldLineRippleAdapter, MDCTextFieldOutlineAdapter, MDCTextFieldRootAdapter} from '@material/textfield/adapter.js';
+import {MDCTextFieldCharacterCounterFoundation} from '@material/textfield/character-counter/foundation.js';
 import MDCTextFieldFoundation from '@material/textfield/foundation.js';
 import {ifDefined} from 'lit-html/directives/if-defined.js';
 
 import {characterCounter, CharacterCounter} from './character-counter/mwc-character-counter-directive.js';
 
+// must be done to get past lit-analyzer checks
+declare global {
+  interface Element {
+    foundation?: MDCLineRippleFoundation|MDCTextFieldCharacterCounterFoundation;
+  }
+}
+
+type ValidityStateProp =
```

Comment:
Mind adding a comment explaining why we need this? And if it's something to do with `readonly` properties, then maybe you can use an `Exclude` similar to this (but not sure I understand this case): https://github.com/material-components/material-components-web-components/blob/a87438e7f9c736f80b6596d05887f5bfcba69d3a/packages/linear-progress/src/mwc-linear-progress-base.ts#L92

Reply:
all of the params of `ValidityState` are readonly. Users are unable to create one, only read it via `HTMLInputElementInstance.validity`. I'm trying to merge the custom validity partial with it and return it as type `ValidityState` for type checking which is why I can't use the keys of `ValidityState` and need my own.